examples/chat/module_0/rust/src/lib.rs

```diff
@@ -14,159 +14,133 @@
-        info!("Create new room {}", req.name);
-        if self.rooms.contains_key(&req.admin_id) {
-            return duplicate_id_err("Admin");
-        }
-        if self.room_to_admin.contains_key(&req.room_id) {
-            return duplicate_id_err("Room");
+        info!("creating room");
+        if self.rooms.contains_key(&req.room_id) {
+            return room_id_duplicate_err();
         }
 
-        // Create a new Node for this room.
-        let (wh, rh) = oak::channel_create().unwrap();
-        oak::node_create("room-config", rh);
-        oak::channel_close(rh.handle);
+        // Create a new Node for this room, and keep the write handle and admin token in the `rooms`
+        // map.
+        self.rooms.insert(req.room_id, Room::new(req.admin_token));
 
-        self.rooms.insert(
-            req.admin_id.clone(),
-            Room {
-                name: req.name,
-                channel: wh,
-            },
-        );
-        self.room_to_admin
-            .insert(req.room_id.clone(), req.admin_id.clone());
-        self.admin_to_room
-            .insert(req.admin_id.clone(), req.room_id.clone());
         Ok(Empty::new())
     }
 
     fn destroy_room(&mut self, req: DestroyRoomRequest) -> grpc::Result<Empty> {
-        if !self.rooms.contains_key(&req.admin_id) {
-            return unknown_id_err("Admin");
+        info!("destroying room");
+        match self.rooms.entry(req.room_id.clone()) {
+            Entry::Occupied(e) => {
+                if e.get().admin_token == req.admin_token {
+                    // Close the only input channel that reaches the per-room Node, which
+                    // will trigger it to terminate.
+                    oak::channel_close(e.get().channel.handle);
+                    e.remove();
+                    Ok(Empty::new())
+                } else {
+                    Err(grpc::build_status(
+                        grpc::Code::PERMISSION_DENIED,
+                        "invalid admin token",
+                    ))
+                }
+            }
+            Entry::Vacant(_) => room_id_not_found_err(),
         }
-        let room = self.rooms.get_mut(&req.admin_id).unwrap();
-        info!("Destroying room {}", room.name);
-
-        // Close the only input channel that reaches the per-room Node, which
-        // will trigger it to terminate.
-        oak::channel_close(room.channel.handle);
-        let user_id = self.admin_to_room.get(&req.admin_id).unwrap().clone();
-        self.admin_to_room.remove(&req.admin_id);
-        self.room_to_admin.remove(&user_id);
-        self.rooms.remove(&req.admin_id);
-        Ok(Empty::new())
     }
 
-    fn join_room(&mut self, req: JoinRoomRequest, mut writer: grpc::ChannelResponseWriter) {
-        let admin_id = match self.room_to_admin.get(&req.room_id) {
+    fn subscribe(&mut self, req: SubscribeRequest, mut writer: grpc::ChannelResponseWriter) {
```

Comment:
The intention here is to clarify that this method is just a convenience to asynchronously receive future messages.

It will not cause policies to change or anything that other participants may care about: participants must have set labels on the messages accordingly from the start, so that all the intended readers are able to get them and subscribe to updates.

e.g. A wants to send a message to B and C, they will tag it as "A or B or C", and from that point on any of A B or C may retrieve it (either in batch, or asynchronously using the subscribe stream). D will not be able to "join" the room; if A B and C want to communicate with D, they will have to create a new room, and attach the label "A or B or C or D" to future messages.

Does it make more sense in this context?